### ISSUE TITLE
investigation with REAL codebase - sj-i/php-profiler

### ISSUE THREAD

**Issue by sasezaki** [struggle-for-php/sfp-psalm-typed-local-variable-plugin#5]
At HEAD `894d821d890e4fcf5ff4d3a8532446fff936cf24`
https://github.com/sj-i/php-profiler/tree/894d821d890e4fcf5ff4d3a8532446fff936cf24

```
$ composer update
$ composer require --dev struggle-for-php/sfp-psalm-typed-local-variable-plugin
$ vendor/bin/psalm-plugin enable struggle-for-php/sfp-psalm-typed-local-variable-plugin
$ ./vendor/bin/psalm
```

analyze by `0.1.1`. would show

```bash
Scanning files...
Analyzing files...

░░░E░░░░░░░░░░░░░E░░░░░E░░E░░░░░E░░EE░░E░░░░░░░░░░░░░░░░░░░E  60 / 118 (50%)
░░░░░░░░░░░░░░░░░░░E░░E░E░░░░░E░E░░░░░░░░░░░░░░░░░░E░░░░E░

ERROR: InvalidScalarTypedLocalVariableIssue - src/Command/Inspector/DaemonCommand.php:161:45 - Type false should be a subtype of true (see https://psalm.dev/000)
                        $searcher_on_read = false;


ERROR: MixedTypeCoercionTypedLocalVariableIssue - src/Inspector/Daemon/Reader/Context/PhpReaderEntryPoint.php:37:33 - Type mixed should be a subtype of PhpProfiler\Inspector\Daemon\Reader\Message\SetSettingsMessage (see https://psalm.dev/000)
        $set_settings_message = yield $channel->receive();


ERROR: MixedTypeCoercionTypedLocalVariableIssue - src/Inspector/Daemon/Reader/Context/PhpReaderEntryPoint.php:41:31 - Type mixed should be a subtype of PhpProfiler\Inspector\Daemon\Reader\Message\AttachMessage (see https://psalm.dev/000)
            $attach_message = yield $channel->receive();


ERROR: MixedTypeCoercionTypedLocalVariableIssue - src/Inspector/Daemon/Searcher/Context/PhpSearcherEntryPoint.php:34:25 - Type mixed should be a subtype of string (see https://psalm.dev/000)
        $target_regex = yield $channel->receive();


ERROR: InvalidScalarTypedLocalVariableIssue - src/Inspector/Settings/DaemonSettings.php:46:24 - Type int(8) should be a subtype of array<array-key, string>|bool|null|string (see https://psalm.dev/000)
            $threads = 8;


ERROR: InvalidScalarTypedLocalVariableIssue - src/Inspector/Settings/DaemonSettings.php:48:20 - Type false|int should be a subtype of array<array-key, string>|bool|null|string (see https://psalm.dev/000)
        $threads = filter_var($threads, FILTER_VALIDATE_INT);


ERROR: InvalidScalarTypedLocalVariableIssue - src/Inspector/Settings/GetTraceSettings.php:43:18 - Type false|int should be a subtype of array<array-key, string>|bool|null|string (see https://psalm.dev/000)
        $depth = filter_var($depth, FILTER_VALIDATE_INT);


ERROR: InvalidScalarTypedLocalVariableIssue - src/Inspector/Settings/TargetProcessSettings.php:46:16 - Type false|int should be a subtype of array<array-key, string>|bool|null|string (see https://psalm.dev/000)
        $pid = filter_var($pid, FILTER_VALIDATE_INT);


ERROR: InvalidScalarTypedLocalVariableIssue - src/Inspector/Settings/TraceLoopSettings.php:50:35 - Type int should be a subtype of array<array-key, string>|bool|null|string (see https://psalm.dev/000)
            $sleep_nano_seconds = self::SLEEP_NANO_SECONDS_DEFAULT;


ERROR: InvalidScalarTypedLocalVariableIssue - src/Inspector/Settings/TraceLoopSettings.php:52:31 - Type false|int should be a subtype of array<array-key, string>|bool|null|string (see https://psalm.dev/000)
        $sleep_nano_seconds = filter_var($sleep_nano_seconds, FILTER_VALIDATE_INT);


ERROR: InvalidScalarTypedLocalVariableIssue - src/Inspector/Settings/TraceLoopSettings.php:61:28 - Type int(10) should be a subtype of array<array-key, string>|bool|null|string (see https://psalm.dev/000)
            $max_retries = self::MAX_RETRY_DEFAULT;


ERROR: InvalidScalarTypedLocalVariableIssue - src/Inspector/Settings/TraceLoopSettings.php:63:24 - Type false|int should be a subtype of array<array-key, string>|bool|null|string (see https://psalm.dev/000)
        $max_retries = filter_var($max_retries, FILTER_VALIDATE_INT);


ERROR: MixedTypeCoercionTypedLocalVariableIssue - src/Lib/Amphp/context-entry.php:26:20 - Type mixed should be a subtype of PhpProfiler\Lib\Amphp\ContextEntryPointInterface (see https://psalm.dev/000)
    $entry_point = $container->make($entry_class);


ERROR: InvalidTypedLocalVariableIssue - src/Lib/Elf/Structure/Elf64/Elf64GnuHashTable.php:98:27 - Type array-key|null should be a subtype of int (see https://psalm.dev/000)
        $last_chain_key = array_key_last($this->chain);


ERROR: InvalidTypedLocalVariableIssue - src/Lib/Loop/AsyncLoopBuilder.php:49:24 - Type PhpProfiler\Lib\Loop\AsyncLoopMiddlewareInterface should be a subtype of null (see https://psalm.dev/000)
            $process = new $loop_class_name(...$parameters);


ERROR: InvalidTypedLocalVariableIssue - src/Lib/Loop/LoopBuilder.php:51:24 - Type PhpProfiler\Lib\Loop\LoopMiddlewareInterface should be a subtype of null (see https://psalm.dev/000)
            $process = new $loop_class_name(...$parameters);


ERROR: InvalidTypedLocalVariableIssue - src/Lib/PhpInternals/ZendTypeReader.php:82:17 - Type FFI\CType should be a subtype of string (see https://psalm.dev/000)
        $type = $ffi->type($type);


ERROR: TypeCoercionTypedLocalVariableIssue - src/Lib/PhpProcessReader/PhpMemoryReader/ExecutorGlobalsReader.php:67:15 - Type FFI\CData should be a subtype of FFI\PhpInternals\zend_executor_globals (see https://psalm.dev/000)
        $eg = $this->readExecutorGlobals($pid, $php_version, $executor_globals_address);


ERROR: TypeCoercionTypedLocalVariableIssue - src/Lib/PhpProcessReader/PhpMemoryReader/ExecutorGlobalsReader.php:70:33 - Type FFI\CData should be a subtype of FFI\PhpInternals\zend_execute_data (see https://psalm.dev/000)
        $current_execute_data = $this->readCurrentExecuteData($pid, $php_version, $eg);


ERROR: TypeCoercionTypedLocalVariableIssue - src/Lib/PhpProcessReader/PhpMemoryReader/ExecutorGlobalsReader.php:73:29 - Type FFI\CData should be a subtype of FFI\PhpInternals\zend_function (see https://psalm.dev/000)
        $current_function = $this->readCurrentFunction($pid, $php_version, $current_execute_data);


ERROR: TypeCoercionTypedLocalVariableIssue - src/Lib/PhpProcessReader/PhpMemoryReader/ExecutorGlobalsReader.php:110:38 - Type FFI\CData should be a subtype of FFI\CPointer (see https://psalm.dev/000)
        $current_execute_data_addr = \FFI::cast('long', $eg->current_execute_data);


ERROR: TypeCoercionTypedLocalVariableIssue - src/Lib/PhpProcessReader/PhpMemoryReader/ExecutorGlobalsReader.php:133:25 - Type FFI\CData should be a subtype of FFI\CPointer (see https://psalm.dev/000)
        $func_pointer = \FFI::cast('long', $current_execute_data->func);


ERROR: TypeCoercionTypedLocalVariableIssue - src/Lib/PhpProcessReader/PhpMemoryReader/ExecutorGlobalsReader.php:158:42 - Type FFI\CData should be a subtype of FFI\CPointer (see https://psalm.dev/000)
        $current_function_name_pointer = \FFI::cast('long', $current_function->common->function_name);


ERROR: TypeCoercionTypedLocalVariableIssue - src/Lib/PhpProcessReader/PhpMemoryReader/ExecutorGlobalsReader.php:167:23 - Type FFI\CData should be a subtype of FFI\PhpInternals\zend_string (see https://psalm.dev/000)
            $string = $this->getTypeReader($php_version)->readAs('zend_string', $current_function_name_zstring);


ERROR: TypeCoercionTypedLocalVariableIssue - src/Lib/PhpProcessReader/PhpMemoryReader/ExecutorGlobalsReader.php:170:47 - Type FFI\CData should be a subtype of FFI\CPointer (see https://psalm.dev/000)
            $current_function_scope_pointer = \FFI::cast('long', $current_function->common->scope);


ERROR: TypeCoercionTypedLocalVariableIssue - src/Lib/PhpProcessReader/PhpMemoryReader/ExecutorGlobalsReader.php:178:32 - Type FFI\CData should be a subtype of FFI\PhpInternals\zend_class_entry (see https://psalm.dev/000)
                $class_entry = $this->getTypeReader($php_version)
                    ->readAs('zend_class_entry', $current_function_class_entry);


ERROR: TypeCoercionTypedLocalVariableIssue - src/Lib/PhpProcessReader/PhpMemoryReader/ExecutorGlobalsReader.php:188:38 - Type FFI\CData should be a subtype of FFI\PhpInternals\zend_string (see https://psalm.dev/000)
                $class_name_string = $this->getTypeReader($php_version)
                    ->readAs('zend_string', $current_class_name_zstring);


ERROR: TypeCoercionTypedLocalVariableIssue - src/Lib/PhpProcessReader/PhpMemoryReader/ExecutorGlobalsReader.php:218:33 - Type FFI\CData should be a subtype of FFI\PhpInternals\zend_string (see https://psalm.dev/000)
            $filename_zstring = $this->getTypeReader($php_version)->readAs('zend_string', $filename_zstring_raw);


ERROR: TypeCoercionTypedLocalVariableIssue - src/Lib/PhpProcessReader/PhpMemoryReader/ExecutorGlobalsReader.php:236:15 - Type FFI\CData should be a subtype of FFI\PhpInternals\zend_executor_globals (see https://psalm.dev/000)
        $eg = $this->readExecutorGlobals($pid, $php_version, $executor_globals_address);


ERROR: TypeCoercionTypedLocalVariableIssue - src/Lib/PhpProcessReader/PhpMemoryReader/ExecutorGlobalsReader.php:239:33 - Type FFI\CData should be a subtype of FFI\PhpInternals\zend_execute_data (see https://psalm.dev/000)
        $current_execute_data = $this->readCurrentExecuteData($pid, $php_version, $eg);


ERROR: TypeCoercionTypedLocalVariableIssue - src/Lib/PhpProcessReader/PhpMemoryReader/ExecutorGlobalsReader.php:244:37 - Type FFI\CData|null should be a subtype of FFI\PhpInternals\zend_execute_data (see https://psalm.dev/000)
            $current_execute_data = $this->readPreviousExecuteData($pid, $php_version, $current_execute_data);


ERROR: TypeCoercionTypedLocalVariableIssue - src/Lib/PhpProcessReader/PhpMemoryReader/ExecutorGlobalsReader.php:254:33 - Type FFI\CData should be a subtype of FFI\PhpInternals\zend_function (see https://psalm.dev/000)
            $current_function = $this->readCurrentFunction($pid, $php_version, $current_execute_data);


ERROR: TypeCoercionTypedLocalVariableIssue - src/Lib/PhpProcessReader/PhpMemoryReader/ExecutorGlobalsReader.php:280:39 - Type FFI\CData should be a subtype of FFI\CPointer (see https://psalm.dev/000)
        $previous_execute_data_addr = \FFI::cast('long', $execute_data->prev_execute_data);


ERROR: TypeCoercionTypedLocalVariableIssue - src/Lib/PhpProcessReader/PhpMemoryReader/ExecutorGlobalsReader.php:304:24 - Type FFI\CData should be a subtype of FFI\CPointer (see https://psalm.dev/000)
        $opline_addr = \FFI::cast('long', $current_execute_data->opline);


ERROR: InvalidTypedLocalVariableIssue - src/Lib/PhpProcessReader/PhpSymbolReaderCreator.php:93:34 - Type int should be a subtype of null (see https://psalm.dev/000)
            $tls_block_address = $tls_finder->findTlsBlock($pid, 1);


ERROR: InvalidTypedLocalVariableIssue - src/Lib/Process/MemoryReader/MemoryReader.php:91:22 - Type FFI\CInteger|int should be a subtype of int (see https://psalm.dev/000)
            $errno = $this->ffi->errno;


ERROR: TypeCoercionTypedLocalVariableIssue - src/Lib/Process/RegisterReader/X64RegisterReader.php:217:17 - Type FFI\CData should be a subtype of FFI\CInteger (see https://psalm.dev/000)
        $zero = $this->ffi->new('long');


ERROR: InvalidTypedLocalVariableIssue - src/Lib/Process/RegisterReader/X64RegisterReader.php:229:22 - Type FFI\CInteger|int should be a subtype of int (see https://psalm.dev/000)
            $errno = $this->ffi->errno;


ERROR: InvalidTypedLocalVariableIssue - src/Lib/Process/RegisterReader/X64RegisterReader.php:239:22 - Type FFI\CInteger|int should be a subtype of int (see https://psalm.dev/000)
            $errno = $this->ffi->errno;


ERROR: InvalidTypedLocalVariableIssue - src/Lib/Process/RegisterReader/X64RegisterReader.php:248:22 - Type FFI\CInteger|int should be a subtype of int (see https://psalm.dev/000)
            $errno = $this->ffi->errno;


------------------------------
40 errors found
------------------------------

Checks took 39.39 seconds and used 116.771MB of memory
Psalm was able to infer types for 99.9694% of the codebase
```

**Comment by sasezaki** [comment]
>  Type false should be a subtype of true

is unexpected result (Bug) .
Other Errors are mostly anticipated result.

**Comment by sasezaki** [comment]
>>   Type false should be a subtype of true

> is unexpected result (Bug) .

fixed with #6
